lib/prelude.dx

```diff
@@ -537,6 +545,30 @@ def sq  [Mul a] (x:a) : a = x * x
 def abs [Add a, Ord a] (x:a) : a = select (x > zero) x (zero - x)
 def mod (x:Int) (y:Int) : Int = rem (y + rem x y) y
 
+instance [Floating a] Floating (n=>a)
+  exp    = map exp
+  exp2   = map exp2
+  log    = map log
+  log2   = map log2
+  log10  = map log10
+  log1p  = map log1p
+  sin    = map sin
+  cos    = map cos
+  tan    = map tan
+  sinh   = map sinh
+  cosh   = map cosh
+  tanh   = map tanh
+  floor  = map floor
+  ceil   = map ceil
+  round  = map round
+  sqrt   = map sqrt
+  pow    = \x y. for i. pow x.i y.i
+  lgamma = map lgamma
+
+def axis1 (x : a => b => c) : b => a => c = for j. for i. x.i.j
+def axis2 (x : a => b => c => d) : c => a => b => d = for k. for i. for j. x.i.j.k
```

Comment:
Rather than "axis", how about renaming these to something like "permute" or "transpose"? Naming ideas not fully polished: just `permute` would be ambiguous, so it'd have to be like `permute{1,2}` or `swapaxes{0to1|0to2}`.

Maybe new APIs can be unlocked when there's a typeclass for index sets. Seems like frst-class index sets would add extra expressive power for defining APIs like axis swapping.

I like the "axis reordering" functions from Swift for TensorFlow, [named `Tensor.transposed(permutation:)`](https://github.com/tensorflow/swift-apis/blob/b7148bc6f5e76bd55e5b11ebd5dd14e42c6ef5a3/Sources/TensorFlow/Operators/Basic.swift#L375-L426):

```swift
extension Tensor {
  /// Returns a transposed tensor, with dimensions permuted in the specified order.
  @inlinable
  @differentiable(wrt: self where Scalar: TensorFlowFloatingPoint)
  public func transposed(permutation: Tensor<Int32>) -> Tensor {
    _Raw.transpose(self, perm: permutation)
  }

  /// Returns a transposed tensor, with dimensions permuted in the specified order.
  @inlinable
  @differentiable(wrt: self where Scalar: TensorFlowFloatingPoint)
  public func transposed(permutation: [Int]) -> Tensor {
    _Raw.transpose(self, perm: permutation)
  }
}
```

Reply:
I thought about something fancy like this. In practice though in Dex I've found I never have more than 3 axes, and 90% of the time I just want to reduce over an inner axis. The names reflect that this is just popping an inner axis.

Reply:
That makes sense! Not over-designing and over-genericizing for now makes sense.